lib/vnet/ipc_credentials.go

```diff
@@ -0,0 +1,207 @@
+}
+
+func newIPCCredentials() (*ipcCredentials, error) {
+	const (
+		// We don't know which clusters will be connected to at this point so
+		// there's no way to fetch the cluster signature_algorithm_suite or unify
+		// suites across multiple root clusters, so just statically use ECDSA
+		// P-256 for these keys.
+		keyAlgo = cryptosuites.ECDSAP256
+		// These certs need to be valid for the full VNet process lifetime, which could be
+		// longer than any individual Teleport session. Going with a month for now.
+		certTTL = 30 * 12 * time.Hour
+	)
+
+	serverCAKey, err := cryptosuites.GenerateKeyWithAlgorithm(keyAlgo)
+	if err != nil {
+		return nil, trace.Wrap(err, "generating key")
+	}
+	clientCAKey, err := cryptosuites.GenerateKeyWithAlgorithm(keyAlgo)
+	if err != nil {
+		return nil, trace.Wrap(err, "generating key")
+	}
+	serverSigner, err := cryptosuites.GenerateKeyWithAlgorithm(keyAlgo)
+	if err != nil {
+		return nil, trace.Wrap(err, "generating key")
+	}
+	clientSigner, err := cryptosuites.GenerateKeyWithAlgorithm(keyAlgo)
+	if err != nil {
+		return nil, trace.Wrap(err, "generating key")
+	}
+
+	serverCAPEM, err := tlsca.GenerateSelfSignedCAWithSigner(serverCAKey, pkix.Name{CommonName: "TeleportVNet Server CA"}, nil /*dnsNames*/, certTTL)
+	if err != nil {
+		return nil, trace.Wrap(err, "generating self-signed CA")
+	}
+	serverCA, err := tlsca.FromCertAndSigner(serverCAPEM, serverCAKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	clientCAPEM, err := tlsca.GenerateSelfSignedCAWithSigner(clientCAKey, pkix.Name{CommonName: "TeleportVNet Client CA"}, nil /*dnsNames*/, certTTL)
+	if err != nil {
+		return nil, trace.Wrap(err, "generating self-signed CA")
+	}
+	clientCA, err := tlsca.FromCertAndSigner(clientCAPEM, clientCAKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	now := time.Now()
+	serverCertPEM, err := serverCA.GenerateCertificate(tlsca.CertificateRequest{
+		PublicKey: serverSigner.Public(),
+		Subject:   pkix.Name{CommonName: "localhost"},
+		DNSNames:  []string{"localhost", "127.0.0.1", "::1"},
+		NotAfter:  now.Add(certTTL),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "generating server TLS certificate")
+	}
+	clientCertPEM, err := clientCA.GenerateCertificate(tlsca.CertificateRequest{
+		PublicKey: clientSigner.Public(),
+		Subject:   pkix.Name{CommonName: "client"},
+		NotAfter:  now.Add(certTTL),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "generating client TLS certificate")
+	}
+
+	return &ipcCredentials{
+		server: credentials{
+			trustedCAPEM: clientCAPEM,
+			certPEM:      serverCertPEM,
+			signer:       serverSigner,
+		},
+		client: credentials{
+			trustedCAPEM: serverCAPEM,
+			certPEM:      clientCertPEM,
+			signer:       clientSigner,
+		},
+	}, nil
+}
+
+func (c *credentials) serverTLSConfig() (*tls.Config, error) {
+	caPool := x509.NewCertPool()
+	if !caPool.AppendCertsFromPEM(c.trustedCAPEM) {
+		return nil, trace.Errorf("Failed to parse CA")
```

Comment:
nit: capitalized error

```suggestion
		return nil, trace.Errorf("parse CA certificate PEM")
```

* https://go.dev/wiki/CodeReviewComments#error-strings